python/ray/autoscaler/v2/instance_manager/instance_storage.py

```diff
@@ -0,0 +1,198 @@
+import logging
+from abc import ABCMeta, abstractmethod
+from dataclasses import dataclass
+from typing import Dict, List, Optional, Set, Tuple
+
+from ray.autoscaler.v2.instance_manager.storage import Storage, StoreStatus
+from ray.core.generated.instance_manager_pb2 import Instance
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class InstanceUpdateEvent:
+    """Notifies the status change of an instance."""
+
+    instance_id: str
+    new_status: int
+
+
+class InstanceUpdatedSuscriber(metaclass=ABCMeta):
+    """Subscribers to instance status changes."""
+
+    @abstractmethod
+    def notify(self, events: List[InstanceUpdateEvent]) -> None:
+        pass
+
+
+class InstanceStorage(object):
+    """Instance storage stores the states of instances in the storage. It also
+    allows users to subscribe to instance status changes to trigger reconciliation
+    with cloud provider."""
+
+    def __init__(
+        self,
+        cluster_id: str,
+        storage: Storage,
+        status_change_subscriber: Optional[InstanceUpdatedSuscriber] = None,
+    ) -> None:
+        super().__init__()
+        self._storage = storage
+        self._cluster_id = cluster_id
+        self._table_name = f"instance_table@{cluster_id}"
+        self._status_change_subscriber = status_change_subscriber
+
+    def batch_upsert_instances(
+        self,
+        updates: List[Instance],
+        expected_storage_version: Optional[int] = None,
+    ) -> StoreStatus:
+        """Upsert instances into the storage. If the instance already exists,
+        it will be updated. Otherwise, it will be inserted. If the
+        expected_storage_version is specified, the update will fail if the
+        current storage version does not match the expected version.
+
+        Note the version of the upserted instances will be set to the current
+        storage version.
+
+        Args:
+            updates: A list of instances to be upserted.
+            expected_storage_version: The expected storage version.
+
+        Returns:
+            StoreStatus: A tuple of (success, storage_version).
+        """
+        mutations = {}
+        version = self._storage.get_version()
+        # handle version mismatch
+        if expected_storage_version and expected_storage_version != version:
```

Comment:
nit. do we need to handle it here? iirc, we already have this logic inside `batch_update`.

Reply:
yeah, potentially it avoid a heavy storage call.